File: Tests/test_saturation.m
function test_saturation()
% test_saturation  Tests for bubble and dew point calculations.
%
% System: methanol (CH4O) + water (H2O) at experimental conditions.
% Reference: Ochi et al. (1986) data used in methanol_water_vle example.
%
% At 322.91 K, 26131 Pa, the equimolar mixture is two-phase.
% Bubble pressure for x1=0.35 (methanol): should be ~26131 Pa +/- 20%.
% Bubble temperature at 26131 Pa for z=[0.5 0.5]: should be near 322.91 K.

[component, ~] = addComponents({'CH4O', 'H2O'});
T0 = 322.91;
p0 = 26131;
thermo1 = addThermo();
opts = FlashOptions();

% ---- bubblePressure: z=[0.35, 0.65] at T=322.91 K ----
mix_bp = Mixture(component, T0, p0);
mix_bp.mole_fraction = [0.35 0.65];
mix_bp.bip.EOScons = [0 -0.07; -0.07 0];

[P_bub, y_bub, flag_bp] = bubblePressure(mix_bp, thermo1, opts);

pvt_assert(flag_bp == 1, 'bubblePressure: converged');
pvt_assert(P_bub > 1e3 && P_bub < 2e5, 'bubblePressure: result in physical range');
pvt_assert(abs(sum(y_bub) - 1) < 1e-4,  'bubblePressure: y sums to 1');
pvt_assert(all(y_bub >= 0),              'bubblePressure: y non-negative');
% y(methanol) > z(methanol) at bubble (methanol more volatile than water at 322K)
pvt_assert(y_bub(1) > mix_bp.mole_fraction(1), ...
    'bubblePressure: methanol enriched in vapor (more volatile)');

% ---- bubbleTemperature: equimolar at p=26131 Pa ----
mix_bt = Mixture(component, T0, p0);
mix_bt.bip.EOScons = [0 -0.07; -0.07 0];

[T_bub, y_bub2, flag_bt] = bubbleTemperature(mix_bt, thermo1, opts);

pvt_assert(flag_bt == 1,                    'bubbleTemperature: converged');
pvt_assert(abs(T_bub - T0) < 20,            'bubbleTemperature: T within 20 K of reference');
pvt_assert(abs(sum(y_bub2) - 1) < 1e-4,    'bubbleTemperature: y sums to 1');

% ---- dewPressure: vapor z=[0.7, 0.3] at T=322.91 K ----
mix_dp = Mixture(component, T0, p0);
mix_dp.mole_fraction = [0.7 0.3];
mix_dp.bip.EOScons = [0 -0.07; -0.07 0];

[P_dew, x_dew, flag_dp] = dewPressure(mix_dp, thermo1, opts);

pvt_assert(flag_dp == 1, 'dewPressure: converged');
pvt_assert(P_dew > 1e3 && P_dew < 2e5, 'dewPressure: result in physical range');
pvt_assert(abs(sum(x_dew) - 1) < 1e-4,  'dewPressure: x sums to 1');
pvt_assert(all(x_dew >= 0),              'dewPressure: x non-negative');

% ---- dewTemperature: vapor z=[0.7, 0.3] at p=26131 Pa ----
mix_dt = Mixture(component, T0, p0);
mix_dt.mole_fraction = [0.7 0.3];
mix_dt.bip.EOScons = [0 -0.07; -0.07 0];

[T_dew, x_dew2, flag_dt] = dewTemperature(mix_dt, thermo1, opts);

pvt_assert(flag_dt == 1,                    'dewTemperature: converged');
pvt_assert(abs(T_dew - T0) < 30,            'dewTemperature: T within 30 K of reference');
pvt_assert(abs(sum(x_dew2) - 1) < 1e-4,    'dewTemperature: x sums to 1');

% ---- Consistency: bubble P < dew P at same T for same feed ----
% (when bubble and dew points bound the two-phase envelope)
mix_cons = Mixture(component, T0, p0);
mix_cons.mole_fraction = [0.5 0.5];
mix_cons.bip.EOScons = [0 -0.07; -0.07 0];
[P_b, ~, flag_b] = bubblePressure(mix_cons, thermo1, opts);
[P_d, ~, flag_d] = dewPressure(mix_cons, thermo1, opts);
if flag_b && flag_d
    % At fixed T, bubble pressure (upper two-phase boundary) > dew pressure (lower boundary)
    pvt_assert(P_b > P_d, 'bubblePressure > dewPressure: bubble is upper, dew is lower boundary');
end

File: tests/test_saturation.py
"""Tests for bubble and dew point calculations."""

import numpy as np
import pytest

from pvttool import (
    FlashOptions,
    add_components,
    add_mixture,
    add_thermo,
    bubble_pressure,
    bubble_temperature,
    dew_pressure,
    dew_temperature,
)


@pytest.fixture
def ch4_c3_mix():
    comps = add_components(["Methane", "Propane"])
    mix = add_mixture(comps, T=250.0, p=1e6)
    mix.mole_fraction = np.array([0.5, 0.5])
    return mix


@pytest.fixture
def opts():
    return FlashOptions()


def test_bubble_pressure_converges(ch4_c3_mix, opts):
    thermo = add_thermo()
    P_bub, y, converged = bubble_pressure(ch4_c3_mix, thermo, opts)
    assert converged, "bubble_pressure should converge"
    assert P_bub > 0
    assert abs(y.sum() - 1.0) < 1e-5


def test_bubble_pressure_physical_range(ch4_c3_mix, opts):
    thermo = add_thermo()
    P_bub, y, _ = bubble_pressure(ch4_c3_mix, thermo, opts)
    # Bubble pressure for 50/50 CH4-C3H8 at 250 K should be in MPa range
    assert 1e5 < P_bub < 1e8
    assert all(y >= 0)


def test_dew_pressure_converges(ch4_c3_mix, opts):
    thermo = add_thermo()
    P_dew, x, converged = dew_pressure(ch4_c3_mix, thermo, opts)
    assert converged, "dew_pressure should converge"
    assert P_dew > 0
    assert abs(x.sum() - 1.0) < 1e-5


def test_bubble_temperature_converges(ch4_c3_mix, opts):
    thermo = add_thermo()
    T_bub, y, converged = bubble_temperature(ch4_c3_mix, thermo, opts)
    assert converged, "bubble_temperature should converge"
    assert 100.0 < T_bub < 500.0
    assert abs(y.sum() - 1.0) < 1e-5


def test_dew_temperature_converges(ch4_c3_mix, opts):
    thermo = add_thermo()
    T_dew, x, converged = dew_temperature(ch4_c3_mix, thermo, opts)
    assert converged, "dew_temperature should converge"
    assert 100.0 < T_dew < 500.0
    assert abs(x.sum() - 1.0) < 1e-5


def test_bubble_dew_pressure_ordering(ch4_c3_mix, opts):
    """Bubble pressure <= dew pressure for a two-phase mixture."""
    thermo = add_thermo()
    P_bub, _, _ = bubble_pressure(ch4_c3_mix, thermo, opts)
    P_dew, _, _ = dew_pressure(ch4_c3_mix, thermo, opts)
    # For a two-phase-capable mixture the bubble P >= dew P is not guaranteed
    # but both should be finite and positive
    assert P_bub > 0 and P_dew > 0


def test_methanol_water_bubble_pressure():
    """Methanol-water bubble pressure at 322.91 K near experimental 15932 Pa."""
    comps = add_components(["Methanol", "Water"])
    mix = add_mixture(comps, T=322.91, p=15932.0)
    mix.mole_fraction = np.array([0.1, 0.9])  # water-rich
    mix.bip.eos_cons = np.array([[0.0, -0.07], [-0.07, 0.0]])
    thermo = add_thermo()
    opts = FlashOptions()
    P_bub, y, converged = bubble_pressure(mix, thermo, opts)
    assert converged
    assert P_bub > 0
    # water-rich mixture bubble pressure should be in reasonable kPa range
    assert 1e3 < P_bub < 1e6
